inst/sampleDB/server_helpers/references/ServerControlsRef.R

```diff
@@ -0,0 +1,474 @@
+		user_file = NULL,
+		user_file_error_annotated = NULL,
+		table = NULL,
+    filters = NULL,
+    dbmap = c(
+      batch_creation_date = "Created", 
+      strain = "Strain", 
+      percentage = "Percentage", 
+      density = "Density", 
+      count = "Count", 
+      bag_name = "BagName",
+      batch = "Batch", 
+      location_name = "Freezer", 
+      level_I = "level_I",
+      level_II = "level_II"
+    ),
+    update.table = c(
+      "strain",
+      "percentage",
+      "density",
+      "bag_name"
+    ),
+    user_selected_rows = NULL,
+    last_selected_row = NULL,
+    rt = NULL
+	)
+
+  error <- reactiveValues(
+    title = "",
+    type = "",
+    message = "",
+    list = NULL
+  )
+
+  observeEvent(rv$error, ignoreInit = TRUE, {
+    message("Running error workflow")
+    df <- error$list
+    modal_size <- "m"
+    if (!is.null(error$type) && error$type == "formatting") {
+      df <- error$list %>%
+        dplyr::rename(
+          Column = column, 
+          Reason = reason,
+          `Triggered By` = trigger
+        ) %>%
+        reactable(.)
+
+      showModal(
+        modalDialog(
+          size = "m",
+          title = error$title,
+          error$message,
+          tags$hr(),
+          renderReactable({ df }),
+          footer = modalButton("Exit")
+        )
+      )
+    } else if (!is.null(error$type) && error$type == "validation") {
+      errors <- unique(names(error$list))
+      errors <- data.frame(errors)
+      colnames(errors) <- "Error"
+      df <- reactable(errors, details = function(index) {
+        data <- error$list[[index]]$Columns
+        htmltools::div(style = "padding: 1rem",
+          reactable(
+            data, 
+            outlined = TRUE, 
+            striped = TRUE,
+            # rownames = TRUE,
+            theme = reactableTheme(
+            headerStyle = list(
+              "&:hover[aria-sort]" = list(background = "hsl(0, 0%, 96%)"),
+              "&[aria-sort='ascending'], &[aria-sort='descending']" = list(background = "hsl(0, 0%, 96%)"),
+              borderColor = "#555"
+            )),
+            defaultColDef = colDef(na = "-", align = "center")
+          )
+        )
+      })
+
+      showModal(
+        modalDialog(
+          size = "l",
+          title = error$title,
+          tags$p("One or more rows had invalid or missing data. See the errors below and expand them to see which rows caused this error."),
+          tags$p("Press the button below to download your file with annotations"),
+          downloadButton("ErrorFileDownload"),
+          tags$hr(),
+          renderReactable({ df }),
+          footer = modalButton("Exit")
+        )
+      )
+  	}
+
+    rv$error <- NULL
+    error$title <- ""
+    error$message <- ""
+    error$type <- ""
+    error$list <- NULL
+	})
+
+  observeEvent(input$InputCreateStrain, ignoreInit = TRUE, {
+
+  	con <- NULL
+ 		tryCatch({
+
+ 			con <- dbConnect(RSQLite::SQLite(), Sys.getenv("SDB_PATH"))
```

Comment:
I feel like in general we should avoid directly accessing the db in the frontend, and instead mediate through the other functions you've written in the R package.